### PR TITLE
Refactor `fossa analyze` entrypoint

### PR DESCRIPTION
**Pull request by scruffystuffs**
# Overview

Closes fossas/team-analysis#828

Implements the new entrypoint model (from fossas/team-analysis#796) for `fossa analyze`.

Notable side changes

- Legacy filters have been removed entirely.
- `FilterCombination` is now strongly typed with `Exclude` or `Include`, and `AllFilters` is specific about accepting one of each.
- The old `App.Fossa.Main` module is dead, and the binary entrypoint refers exclusively to the new one.  As a result, **THIS PR IS NOT SAFE TO BE MERGED DIRECTLY INTO MASTER!  DON'T MERGE THIS INTO MASTER!  DON'T APPROVE THIS PR IF IT'S MERGING INTO MASTER!**
- More `MonadIO` constraints have been removed.

## Acceptance criteria

- The business-logic-module of `fossa analyze` sees configuration as a immutable fact, rather than actively participating in the configuration process.
- The configuration module of `fossa analyze` only exposes types which are either opaque or representative of the final configuration "fact".
- Config errors are collected, and reported together.  i.e.: an invalid basedir and a missing API key (assuming no `--output` flag) are both reported together, rather than aborting at the first error.  This is done using `Validation`, from the `validation-selective` package.
- No semantic changes except for:
  - Missing subcommands (`fossa test`, `fossa report`, etc. those are follow-up tickets)
  - Monorepo is temporarily broken.
  - Config file accept/ignore/warn/abort semantics are change via [this slack conversation](https://teamfossa.slack.com/archives/G01DUC2JW3B/p1639173134260800)

## Testing plan

Run `fossa analyze`, without any monorepo options.

## Risks

- This is not safe to merge into master.
- It's very possible that the configuration rewrite has mistakes that ignore some config values from certain places.  Reviewers, please pay close attention to those configuration resolving functions.
- This PR lays out a massive change to our CLI frontend, and its conclusion will become the basis for the future of the CLI and configuration.  We should not merge/approve it until we are convinced that this model is the best way forward for the CLI.

## References

Task Ticket: fossas/team-analysis#828
Project Ticket: fossas/team-analysis#796

## Checklist

- [x] ~I added tests for this PR's change (or confirmed tests are not viable).~
- [x] ~If this PR introduced a user-visible change, I added documentation into `docs/`.~
- [x] ~I updated `Changelog.md` if this change is externally facing. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.~
- [x] ~I updated `*schema.json` if I have made changes for `.fossa.yml`, `fossa-deps.{json, yaml, yml}`. You may also need to update these if you have added/removed new dependency (e.g. pip) or analysis target type (e.g. poetry).~
- [x] I linked this PR to any referenced GitHub issues, if they exist.

### CODE REVIEW

**Review by kitified — APPROVED** [review]
I like where this is headed! Have a couple smaller comments but overall very happy with this design. It's clear, it makes sense, and it is more unidirectional than the config mishmash we had before. Nice work!

File: src/App/NewFossa/Config/Common.hs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  { optDebug :: Bool
+  , optBaseUrl :: Maybe URI
+  , optProjectName :: Maybe Text
+  , optProjectRevision :: Maybe Text
+  , optAPIKey :: Maybe Text
+  , optConfig :: Maybe FilePath
+  }
+
+globalOpts :: Parser GlobalOpts
+globalOpts =
```
[the remark this answers is not in the document]
Comment (scruffystuffs) [reply src/App/NewFossa/Config/Common.hs:83]: I think I'll go with `CommonOpts`, matches the module name it lives in.  Moved this task to #756

File: spectrometer.cabal [hunk 2]
```diff
@@ -178,6 +178,12 @@ library
     App.Fossa.VSI.IAT.Types
     App.Fossa.VSI.Types
     App.Fossa.VSIDeps
+    App.NewFossa.Config.Analyze
+    App.NewFossa.Config.Common
+    App.NewFossa.ConfigFile
```
Comment (meghfossa) [inline spectrometer.cabal:183]: [optional] [nit] prefer `ConfigFile` and `EnvironmentVars ` modules in config namespace

**Review by meghfossa — APPROVED** [review]
LGTM - left one nit. 

I did not perform test on my machine, leaving that for PR targeting master.

File: spectrometer.cabal (hunk 2 above)
Comment (scruffystuffs) [reply spectrometer.cabal:183]: Moved to #756

File: src/App/Fossa/Analyze.hs [hunk 3]
```diff
@@ -332,33 +286,40 @@ analyze ::
...
   -- additional source units are built outside the standard strategy flow, because they either
   -- require additional information (eg API credentials), or they return additional information (eg user deps).
   vsiResults <-
-    Diag.context "analyze-vsi" . runStickyLogger SevInfo . runFinally $
-      analyzeVSI modeVSIAnalysis apiOpts basedir override filters modeVSISkipResolution
+    Diag.context "analyze-vsi" . runStickyLogger SevInfo . runFinally $ do
+      let shouldRunVSI = fromFlag Config.VSIAnalysis $ Config.vsiAnalysisEnabled $ Config.vsiOptions cfg
+      case (shouldRunVSI, apiOpts) of
+        (True, Just apiOpts') -> analyzeVSI apiOpts' basedir revision filters skipResolutionSet
+        _ -> pure Nothing
```
Comment (scruffystuffs) [inline src/App/Fossa/Analyze.hs:317]: needs re-review (@kitified)

**Review by kitified — APPROVED** [review]
Flagged block and associated `analyzeVSI` function lgtm 👍

File: src/App/Fossa/Analyze.hs [hunk 4]
```diff
@@ -427,24 +380,29 @@ analyzeVSI VSIAnalysisEnabled (Just apiOpts) dir override filters skipResolving
       traverse_ (logInfo . pretty . VSI.renderLocator) skippedLocators
     else pure ()
 
-  revision <- inferProjectRevision dir override
   results <- analyzeVSIDeps dir revision apiOpts filters skipResolving
   pure $ Just results
-analyzeVSI _ _ _ _ _ _ = pure Nothing
```
Comment (scruffystuffs) [inline src/App/Fossa/Analyze.hs:385]: needs re-review (@kitified)

**Pull request merged**